### PR TITLE
adding documentation for mid-game reconnection [MTT-3469]

### PR DESCRIPTION
**Pull request by LPLafontaineB**
**Select the type of change:**

- [ ] Small Changes - Typos, formatting, slight revisions
- [x] New Content - New features, sections, pages, tutorials
- [ ] Site and Tools - Updates, maintenance, and new packages for the site and Docusaurus

**Purpose of the Pull Request:**

This new content describes what users have to keep in mind to allow clients to reconnect mid-game.

**Issue Number:** 

[MTT-3469](https://jira.unity3d.com/browse/MTT-3469)

### CODE REVIEW

File: docs/advanced-topics/reconnecting-mid-game.md
Comment (SamuelBellomo) [inline docs/advanced-topics/reconnecting-mid-game.md:1]: I don't see any link to our auto reconnect code in boss room?

Comment (LPLafontaineB) [reply docs/advanced-topics/reconnecting-mid-game.md:1]: (currently pointing to develop since this feature is not in a released version of BR yet. When the release happens, we should update this link to point to main instead)

**Review by SamuelBellomo — APPROVED** [review]
There's some sentences that seem "french", but I'll leave this to the pros @s-omeilia-unity  :D

File: docs/advanced-topics/reconnecting-mid-game.md
Comment (NoelStephensUnity) [inline docs/advanced-topics/reconnecting-mid-game.md:6]: "...want to give your players the ability to reconnect to the game."

File: docs/advanced-topics/reconnecting-mid-game.md
Comment (NoelStephensUnity) [inline docs/advanced-topics/reconnecting-mid-game.md:8]: "The following are some considerations to keep in mind when providing mid-game reconnection capabilities to your players:"

File: docs/advanced-topics/reconnecting-mid-game.md
Comment (NoelStephensUnity) [inline docs/advanced-topics/reconnecting-mid-game.md:9]: "In order to preserve a player's in-game state when they are disconnected, you will need to implement a mechanism to keep the player's in-game state in memory and restore that state upon the client reconnecting (see..."

File: docs/advanced-topics/reconnecting-mid-game.md
Comment (NoelStephensUnity) [inline docs/advanced-topics/reconnecting-mid-game.md:10]: - General Guidelines for when a client is disconnected:
  - Preserve the client's current in-game state on both the client and server side
    - This will provide you with any state deltas between the client and server that might have occurred in the time span between when the client was disconnected and when the server detects the client was disconnected.
  - Restore the client's scene hierarchy state as it was just prior to the client having connected for the first time.
    - Typically unloading any additive scenes loaded during the game-session or returning to a "single mode" loaded lobby will accomplish this
- General Guidelines for reconnecting a client:
  - Reconnect the client and wait for the client to be synchronized
      - Once synchronized, use the saved client state to restore the client's state as the last part of synchronization.
- When scene management is enabled:
  - Wait until the client's initial connection synchronization event is triggered before applying the reconnected client's game state.   
  There are three ways to monitor when a client is synchronized:
    - Subscribing to the `NetworkSceneManager.OnSynchronizeComplete` event.
    - Subscribing to the `NetworkSceneManager.OnSceneEvent` and checking for the `SceneEvent.SceneEventType`  to equal `SceneEventType.SynchronizeComplete`.
    - Subscribing to the `NetworkManager.OnClientConnectedCallback`
      - _`The NetworkSceneManager` invokes this when a client's synchronization is complete_
- When scene management is disabled:
  - The same principles apply when scene management is disabled with the exception that you might have additional unique scenarios to take into consideration depending upon the custom implementation.

File: docs/advanced-topics/reconnecting-mid-game.md
Comment (NoelStephensUnity) [inline docs/advanced-topics/reconnecting-mid-game.md:16]: Automatically reconnecting a client that has lost its connection unexpectedly improves the player experience.
Guidelines for reconnecting a client automatically:
- Determine what will initiate the reconnection process.
  - For example, it could be the `NetworkManager.OnClientDisconnectCallback` callback or it could be some other unique event specific to your product.
- Wait for `NetworkManager` to properly shut down before attempting to reconnect. You can use the `NetworkManager.ShutdownInProgress` property for this.
  - Typically it takes at least 1 frame for this to happen
- Add additional checks to assure  your automatic client reconnection solution it not triggered under the wrong conditions.
  - For example, if a client quits a game purposefully, if the server shuts down normally as the game session ends, if a client gets kicked from a game or if they get denied during connection approval you do not want the client to automatically reconnect.

**Review by NoelStephensUnity — COMMENTED** [review]
I left some feedback on potential adjustments/additions.

File: docs/advanced-topics/reconnecting-mid-game.md
Comment (LPLafontaineB) [reply docs/advanced-topics/reconnecting-mid-game.md:10]: @NoelStephensUnity  I have questions about some of these points:

> * General Guidelines for when a client is disconnected:
>      * Preserve the client's current in-game state on both the client and server side
>          * This will provide you with any state deltas between the client and server that might have occurred in the time span between when the client was disconnected and when the server detects the client was disconnected.

Why do we need those deltas? Shouldn't we only consider the state server-side when it detects the client disconnect?

>      * Restore the client's scene hierarchy state as it was just prior to the client having connected for the first time.
>         * Typically unloading any additive scenes loaded during the game-session or returning to a "single mode" loaded lobby will accomplish this

I can see cases where users might want to keep the scene hierarchy. For example, that is what is done in Boss Room during the automatic reconnect process. Should we mention it here?

**Comment by s-omeilia-unity** [comment]
@LPLafontaineB @SamuelBellomo @NoelStephensUnity 

Made some edits for clarification. I left a few comments inside the doc while I was working through it for further clarification. Let me know if I misinterpreted any of your content when I did my pass.

Thanks y'all!

File: docs/advanced-topics/reconnecting-mid-game.md
Comment (LPLafontaineB) [inline docs/advanced-topics/reconnecting-mid-game.md:13]: This is really about one specific case where there are some effects that can be undesired in some cases:
If the active scene on the client is the same as the active scene on the server, the client will skip loading the server scene in single mode, but not additive scenes.
This means, that if the client was connected and over the course of the connection, some additive scenes were loaded by the server, the client could still have them loaded when disconnecting (if they don't switch scene when that happens). In that case, when reconnecting, if the server also has the same additive scenes currently loaded, the client would load every additive scene a second time.
But this wasn't very clear in my initial draft, so that's why @NoelStephensUnity suggested these changes : https://github.com/Unity-Technologies/com.unity.multiplayer.docs/pull/614#discussion_r876071897

File: docs/advanced-topics/reconnecting-mid-game.md
Comment (LPLafontaineB) [inline docs/advanced-topics/reconnecting-mid-game.md:25]: They're not really steps, more like guidelines to follow, so the order does not matter

File: docs/advanced-topics/reconnecting-mid-game.md
Comment (LPLafontaineB) [inline docs/advanced-topics/reconnecting-mid-game.md:26]: Yes, the user must define this

File: docs/advanced-topics/reconnecting-mid-game.md
Comment (NoelStephensUnity) [reply docs/advanced-topics/reconnecting-mid-game.md]: This is really more of a suggestion... there could be some unique aspects to the users netcode assets that might require them to do some of the "delta checking/adjusting" on the client side.  
Another example might be a game where a player disconnects with an internally stored time that contains their data and perhaps the last "local player health update"...if the player drops and the server doesn't detect this for several seconds then something could have impacted the disconnected player's health (i.e. player couldn't do anything about it) and so the user might add some additional code when it receives what the server "thinks" is the player's health but then the player's client responds with the last health update information and some verification key (delivered when updating)... the server validates the key, compares the timestamp of when it was sent (could be part of the key), and if it is valid restores the player's health server side.

I was just trying to keep a broad stroke brush (i.e. save state when disconnected in all places pertinent...could be only server...could be both server and client).

However, since it was only a suggestion... feel free to delete/remove anything that doesn't make 100% sense to have here.

Comment (LPLafontaineB) [reply docs/advanced-topics/reconnecting-mid-game.md]: I see. I think in that case that it would make more sense to have it in the session management doc page

**Review by s-omeilia-unity — APPROVED** [review]
Made a final pass on the updates you made @LPLafontaineB 

I'm set with my review.

**Pull request merged**
